KR registration and log in

kjreczek opened the pull request:
I made views for registration, login and password reset. I used material-ui for this. 
I tried to do a storybook, but it failed to display these components every time i tried, and I finally gave up.

Logic is done for registration and login. If you want to see it work mount it in App or wherever you want.

robert-czwartosz commented on frontend/src/components/LogIn/LogIn.stories.tsx:
```diff
@@ -0,0 +1,15 @@
+import React from 'react';
+import { Story, Meta } from '@storybook/react/types-6-0';
+import LogIn , { LogInProps } from './LogIn';
```
If you want to configure storybook, add folowing imports:
```
import { Provider } from 'react-redux';
import { store } from '../../app/store';
```

kjreczek replied:
thanks! 💯

robert-czwartosz commented on frontend/src/components/LogIn/LogIn.stories.tsx:
```diff
@@ -0,0 +1,15 @@
+  component: LogIn,
+} as Meta;
+
+const Template: Story<LogInProps> = (args) => <LogIn {...args} />;
```
```
const Template: Story<StyledTextFieldProps> = (args) => (
        <Provider store={store}>
          <LogInProps {...args} />
        </Provider>
);
```

robert-czwartosz commented on frontend/src/components/LogIn/LogIn.stories.tsx:
```diff
@@ -0,0 +1,15 @@
+
+const Template: Story<LogInProps> = (args) => <LogIn {...args} />;
+
+export const FirstInput = Template.bind({});
+FirstInput.args = {
+  label: 'Write something: ',
+};
```
...and add some props if it's needed

robert-czwartosz commented on frontend/src/components/Registration/RegistrationSlice.ts:
```diff
@@ -0,0 +1,25 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+import { AppThunk, RootState } from '../../app/store';
+
+interface RegistrationState {
+  
+}
+
+const initialState: RegistrationState = {
+  
+};
+
+export const registrationSlice = createSlice({
+  name: 'registration',
+  initialState,
+  reducers: {
+
+  },
+});
+
+export const { } = registrationSlice.actions;
+
+// if you want, add selectors here, change the one below, remember to register reducer in store.ts
+// export const selectRegistration = (state: RootState) => state.registration.value;
+
+export default registrationSlice.reducer;
```
I suggest to remove this file

kjreczek replied:
I forgot about that 🤦‍♂️

robert-czwartosz commented on frontend/src/components/ResetPassword/ResetPasswordSlice.ts:
```diff
@@ -0,0 +1,25 @@
+import { createSlice, PayloadAction } from '@reduxjs/toolkit';
+import { AppThunk, RootState } from '../../app/store';
+
+interface ResetPasswordState {
+  
+}
+
+const initialState: ResetPasswordState = {
+  
+};
+
+export const resetPasswordSlice = createSlice({
+  name: 'resetPassword',
+  initialState,
+  reducers: {
+
+  },
+});
+
+export const { } = resetPasswordSlice.actions;
+
+// if you want, add selectors here, change the one below, remember to register reducer in store.ts
+// export const selectResetPassword = (state: RootState) => state.resetPassword.value;
+
+export default resetPasswordSlice.reducer;
```
I suggest to remove this file

kjreczek replied:
done 👍

robert-czwartosz commented on package.json:
```diff
@@ -73,8 +73,10 @@
     "testEnvironment": "node"
   },
   "dependencies": {
+    "axios": "^0.21.1",
```
I think that axios package is not needed for backend app.

kjreczek replied:
my mistake, done 👍

robert-czwartosz commented on frontend/src/components/LogIn/LogIn.tsx:
```diff
@@ -0,0 +1,97 @@
+  const handleSignInClick = async () => {
+    const service = new BaseService;
+    try {
+      const response = await service.post('login', { email, password})
+      setFormError('');
+      const token: any = response?.headers?.['x-auth-token'];
+      localStorage.setItem('token', token);
+      dispatch(setLoggedInUserId(response?.data?._id || ''));
```
In my opinion, token and role should be saved in redux store.

robert-czwartosz commented on frontend/src/components/LogIn/LogIn.tsx:
```diff
@@ -0,0 +1,97 @@
+  const [formError, setFormError] = useState('');
+
+  const handleSignInClick = async () => {
+    const service = new BaseService;
```
`const service = new BaseService();`

kjreczek replied:
👍

kjreczek commented on frontend/src/App.test.tsx:
```diff
@@ -5,11 +5,11 @@ import { store } from './app/store';
-  const { getByText } = render(
-    <Provider store={store}>
-      <App />
-    </Provider>
-  );
+  // const { getByText } = render(
+  //   <Provider store={store}>
+  //     <App />
+  //   </Provider>
+  // );
 
-  expect(getByText(/learn/i)).toBeInTheDocument();
+  // expect(getByText(/learn/i)).toBeInTheDocument();
 });
```
right, done 👍

kjreczek commented on frontend/src/components/Registration/Registration.tsx:
```diff
@@ -0,0 +1,129 @@
+import React, { useState } from 'react';
+
+import Button from '@material-ui/core/Button';
+import CssBaseline from '@material-ui/core/CssBaseline';
+import Link from '@material-ui/core/Link';
+import Grid from '@material-ui/core/Grid';
+import Typography from '@material-ui/core/Typography';
+import Container from '@material-ui/core/Container';
+import FormHelperText from '@material-ui/core/FormHelperText';
```
👍

kjreczek commented on frontend/src/components/Registration/Registration.test.tsx:
```diff
@@ -0,0 +1,79 @@
+         </Provider>
+      );
+
+      const fnameInput = await screen.getByTestId('r-fname');
+      const lnameInput = await screen.getByTestId('r-lname');
+      const emailinput = await screen.getByTestId('r-email');
+      const passwordInput = await screen.getByTestId('r-password');
+      const cpasswordInput = await screen.getByTestId('r-cpassword');
+      const button = await screen.getByTestId('r-button');
```
👍

kjreczek commented on frontend/src/components/Registration/Registration.test.tsx:
```diff
@@ -0,0 +1,79 @@
+         </Provider>
+      );
+
+      const inputFirstNameDiv = await screen.getByTestId('r-fname')
```
👍

robert-czwartosz commented on frontend/src/app/store.ts:
```diff
@@ -1,10 +1,12 @@
 export const store = configureStore({
   reducer: {
     counter: counterReducer,
+    login: loginReducer,
```
On yesterday meeting we've decided to store token, user id and role in local storage.
... so login reducer looks be unnecessary.

kjreczek replied:
👍

kjreczek commented:
I added route to login, registration and resetpassword, also added a condition that it should display the login view for users whose data (token and i add id and type) are not in LS and for those whose data are, it should display /home.
I create a function to app/utils that returns data from LS, this will probably be useful and I changed the error display.

robert-czwartosz commented on frontend/src/components/Registration/Registration.test.tsx:
```diff
@@ -0,0 +1,83 @@
+
+describe('Registration', () => {
+   jest.setTimeout(7000);
+   // it('renders without error', () => {
+   //    render(
+   //       <Provider store={store}>
+   //          <Registration />)
+   //       </Provider>
+   //    );
+
+   //    const fnameInput = screen.getByTestId('r-fname');
+   //    const lnameInput = screen.getByTestId('r-lname');
+   //    const emailinput = screen.getByTestId('r-email');
+   //    const passwordInput = screen.getByTestId('r-password');
+   //    const cpasswordInput = screen.getByTestId('r-cpassword');
+   //    const button = screen.getByTestId('r-button');
+
+   //    expect(fnameInput).toHaveTextContent('Name');
+   //    expect(lnameInput).toHaveTextContent('Surname');
+   //    expect(emailinput).toHaveTextContent('Email Address');
+   //    expect(passwordInput).toBeInTheDocument();
+   //    expect(cpasswordInput).toBeInTheDocument();
+   //    expect(button ).toBeInTheDocument();
+   // });
```
Leaving commented code doesn't look good...

kjreczek replied:
When I was working on the tests it was throwing me a bug at one point and I didn't know which test it was throwing for, so I commented code on the good one... and forgot to undo it. My mistake. 🙈

robert-czwartosz reviewed (commented):
There is issue with some commented code.
Generally, great job! :+1: 
Especially, awesome error handling.

robert-czwartosz commented on frontend/.env:
```diff
@@ -1 +1,3 @@
\ No newline at end of file
+SKIP_PREFLIGHT_CHECK=true
+
+REACT_APP_API_URL=http://localhost:5000
```
I think better would be setting
`REACT_APP_API_URL=https://coders-camp-organization-app.herokuapp.com`

kjreczek replied:
I changed url 👍

fszwedo commented on frontend/src/components/LogIn/LogIn.tsx:
```diff
@@ -0,0 +1,135 @@
+import useStyles from './LogIn.style';
+import BaseService from '../../app/baseService';
+
+export interface LogInProps {
+
+};
```
seems unnecesary (unless you expect there might be some changes in the future here)

kjreczek replied:
For now, I would leave it as it is, because if we want to change something, it will be at hand, and it does not bother whether it is or not.

fszwedo commented on frontend/src/components/Registration/Registration.tsx:
```diff
@@ -0,0 +1,159 @@
+import BaseService from '../../app/baseService';
+import useStyles from './Registration.style';
+
+export interface RegistrationProps {
+  
+};
```
seems unnecesary (unless you expect there might be some changes in the future here)

kjreczek replied:
👍

fszwedo commented on frontend/src/components/ResetPassword/ResetPassword.tsx:
```diff
@@ -0,0 +1,70 @@
+
+import StyledTextField from '../StyledTextField'
+
+export interface ResetPasswordProps {
+
+};
```
seems unnecesary (unless you expect there might be some changes in the future here)

kjreczek replied:
I don't know how it will be done in the future, I just made a component by the way (looks similary to login and registration :P), someone else will take care of the logic, so I'll leave it as it is.

fszwedo commented on frontend/src/components/StyledTextField/StyledTextField.style.ts:
```diff
@@ -0,0 +1,20 @@
+    color : '#FFF !important'
+  },
+
+  cssFocused: { },
```
seems unnecessary

kjreczek replied:
I know it looks weird, but it has to stay that way otherwise styles in material-ui don't work.

fszwedo reviewed (approved):
there are just some probably unnecesary interfaces, apart from that all looks awesome

all tested and works as expected (except for pw reset but I understand powering it with api connection is to be done in scope of a separate task)

fszwedo commented:
@kjreczek please look at my comments, once you confirm the things can be removed or should stay - feel free to merge :)
